lib/CL/devices/cuda/pocl-cuda.c

```diff
@@ -0,0 +1,679 @@
+/* pocl-cuda.c - driver for CUDA devices
+
+   Copyright (c) 2016 James Price / University of Bristol
+
+   Permission is hereby granted, free of charge, to any person obtaining a copy
+   of this software and associated documentation files (the "Software"), to
+   deal
+   in the Software without restriction, including without limitation the rights
+   to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+   copies of the Software, and to permit persons to whom the Software is
+   furnished to do so, subject to the following conditions:
+
+   The above copyright notice and this permission notice shall be included in
+   all copies or substantial portions of the Software.
+
+   THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+   IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+   FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+   AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+   LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+   FROM,
+   OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+   THE SOFTWARE.
+*/
+
+#include "config.h"
+
+#include "common.h"
+#include "devices.h"
+#include "pocl-cuda.h"
+#include "pocl-ptx-gen.h"
+#include "pocl_cache.h"
+#include "pocl_file_util.h"
+#include "pocl_runtime_config.h"
+#include "pocl_util.h"
+
+#include <string.h>
+
+#include <cuda.h>
+
+typedef struct pocl_cuda_device_data_s
+{
+  CUdevice device;
+  CUcontext context;
+} pocl_cuda_device_data_t;
+
+static void
+pocl_cuda_abort_on_error (CUresult result, unsigned line, const char *func,
+                          const char *code, const char *api)
+{
+  if (result != CUDA_SUCCESS)
+    {
+      const char *err_name;
+      const char *err_string;
+      cuGetErrorName (result, &err_name);
+      cuGetErrorString (result, &err_string);
+      POCL_MSG_PRINT2 (func, line, "Error during %s\n", api);
+      POCL_ABORT ("%s: %s\n", err_name, err_string);
+    }
+}
+
+#define CUDA_CHECK(result, api)                                               \
+  pocl_cuda_abort_on_error (result, __LINE__, __FUNCTION__, #result, api);
+
+void
+pocl_cuda_init_device_ops (struct pocl_device_ops *ops)
+{
+  pocl_basic_init_device_ops (ops);
+
+  ops->device_name = "CUDA";
+  ops->init_device_infos = pocl_cuda_init_device_infos;
+  ops->probe = pocl_cuda_probe;
+  ops->uninit = pocl_cuda_uninit;
+  ops->init = pocl_cuda_init;
+  ops->alloc_mem_obj = pocl_cuda_alloc_mem_obj;
+  ops->free = pocl_cuda_free;
+  ops->compile_kernel = pocl_cuda_compile_kernel;
+  ops->read = pocl_cuda_read;
+  ops->read_rect = pocl_cuda_read_rect;
+  ops->write = pocl_cuda_write;
+  ops->write_rect = pocl_cuda_write_rect;
+  ops->copy = pocl_cuda_copy;
+  ops->copy_rect = pocl_cuda_copy_rect;
+  // ops->get_timer_value = pocl_cuda_get_timer_value;
```

Comment:
Please avoid commented out code and also C++ style comments. If you comment out code, #if 0 is the recommended option.